### ISSUE TITLE
Dynamo uses CONSTANT_MATCH guards for string inputs

### ISSUE THREAD

**Issue by jiyuanq** [pytorch/pytorch#93923]
### 🐛 Describe the bug

There are cases when we define pytorch modules that take strings or list[str] as inputs. For example, torchtext.vocab which does vocab lookup for string tokens, and modules that use torchtext.vocab underlying.

With jit script this works fine and string inputs are treated as variable input, but with dynamo they are somehow treated as constants, leading to recompilation every time the values change.

I'm not sure if this is expected behavior. If yes, this can be quite more restrictive than torchscript, especially for model export

### Error logs

![截屏2023-02-02 下午5 09 45](https://user-images.githubusercontent.com/90181230/216287269-fc42e8f0-eb2b-4fd9-801d-38f768d758d4.png)

The compiled graph should not have `[1, 0]` as constants

### Versions

PyTorch version: 2.0.0.dev20230201+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.10.130-118.517.amzn2.x86_64-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] pytorch-lightning==1.5.10
[pip3] torch==2.0.0.dev20230201+cpu
[pip3] torch-tb-profiler==0.4.0
[pip3] torchdata==0.6.0.dev20230201
[pip3] torchmetrics==0.9.3
[pip3] torchtext==0.15.0.dev20230201+cpu
[conda] nomkl                     1.0                  h5ca1d4c_0    conda-forge
[conda] numpy                     1.22.4          py310h4ef5377_0    conda-forge
[conda] pytorch-lightning         1.5.10                   pypi_0    pypi
[conda] torch                     2.0.0.dev20230201+cpu          pypi_0    pypi
[conda] torch-tb-profiler         0.4.0                    pypi_0    pypi
[conda] torchdata                 0.6.0.dev20230201          pypi_0    pypi
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchtext                 0.15.0.dev20230201+cpu          pypi_0    pypi

**Comment by gchanan** [comment]
I was able to reproduce this.  Putting as triage review because it's not clear what the desired behavior is.

**Comment by gchanan** [comment]
Here is the raw text:

```
import torch
import torchtext
from collections import Counter, OrderedDict
from typing import List
import torch._dynamo as dynamo

class TokenToLongTensor(torch.nn.Module):
    def __init__(self, vocab):
        super().__init__()
        self.vocab = vocab

    def forward(self, token : List[str]) -> torch.Tensor:
        return torch.tensor(self.vocab(token))

counter = Counter(["a", "a", "b", "b", "b"])
sorted_by_freq_tuples = sorted(counter.items(), key=lambda x: x[1], reverse=True)
ordered_dict = OrderedDict(sorted_by_freq_tuples)
v1 = torchtext.vocab.vocab(ordered_dict)
m=TokenToLongTensor(v1)

print(m(['a', 'b']))

explanation, out_Guards, graphs, ops_per_graph, _, reason = dynamo.explain(m, ['a', 'b'])

print(explanation)

print(graphs[0].graph)

c=torch.compile(m)
```

**Comment by jiyuanq** [comment]
> I was able to reproduce this. Putting as triage review because it's not clear what the desired behavior is.

Thanks! I think it's counter-intuitive that the module gets recompiled everytime the input strings change.

As a user who have been using jit script quite a lot in my projects to compile string data processing logic into torchscript modules for the speed up and portability, I hope I can still leverage dynamo for speeding up my models.

**Comment by ezyang** [comment]
PT2 stack does not optimize string computation (and similarly, export will not catch it.) If you want it to be captured, you will need to convert the strings into tensors (maybe using NestedTensor for efficiency.)

**Comment by jiyuanq** [comment]
> PT2 stack does not optimize string computation (and similarly, export will not catch it.) If you want it to be captured, you will need to convert the strings into tensors (maybe using NestedTensor for efficiency.)

Thanks for the reply! @ezyang 
When you say "convert the strings into tensors", which of the following do you mean?

1. Apply vocab lookup or hashing trick to convert strings to ints
2. Store bytes of the strings in byte tensors

(1) is exactly what we are trying to achieve with string processing modules, `torchtext.vocab` is a good example.

(2) I've thought about but realized it won't work well. If we turn strings to byte tensors we'll have to implement hash function or dictionary lookup function that takes byte tensors rather than strings as input, which is a roundabout way to do a straightforward thing.

I think string computation modules like `torchtext.vocab` exists today because it makes sense for logic like this to be bundled with the model itself, and torchscript makes it easily portable.

If pt2 does not capture string computation, what is the recommended way to adapt models using such modules?

**Comment by ezyang** [comment]
Oh sorry, I think I misunderstood your problem. You should just disable dynamo on these, using torch._dynamo.disable, as we don't expect to be able to compile this. Once you have tensors, which is what your code does, you can use PT2 as usual.

PT2 seamlessly interoperates with regular Python code, I guess the unfortunate thing is that we don't realize we shouldn't try to compile this code.

**Comment by jiyuanq** [comment]
> Oh sorry, I think I misunderstood your problem. You should just disable dynamo on these, using torch._dynamo.disable, as we don't expect to be able to compile this. Once you have tensors, which is what your code does, you can use PT2 as usual.
> 
> PT2 seamlessly interoperates with regular Python code, I guess the unfortunate thing is that we don't realize we shouldn't try to compile this code.

I see. If this is the case, is it fair to say that the upcoming export functionality for pt2 won't be able to bundle string data preprocessing logic into the model the same way as torchscript does?

**Comment by ezyang** [comment]
That's right, if you need to export code processing Python strings you will need some other approach (TS being one possibility).

**Comment by jiyuanq** [comment]
@ezyang thanks! 
A few follow up questions:
1. what's the long term plan on that front. As in, is there a plan to bridge this gap between pt2 dynamo model export and torchscript model export, for users who may want both speedup from pt2 and portability for string feature preprocessing?
2. Is torchscript going to get continued support and improvement over the longer term as dynamo gets more usage?
3. Are there other alternatives for portable data preprocessing use cases? torcharrow seems promising but it's still in beta and the documentation is sparse

**Comment by ezyang** [comment]
I don't know, unfortunately. @raziel, what do you think our stance on string preprocessing (e.g., tokenization) is going to be in the executorch world

**Comment by raziel** [comment]
Yeah. The current focus is on being able to trace them but as constants. Main use case that has come up for supporting string is as people use them within their modules to store some configuration (constant) values.

**Comment by jiyuanq** [comment]
@raziel Thanks for the comprehensive reply!
As a user I do hope pytorch provides a clean and performant preprocessing solution, especially for string data, that is also portable for serving. 
To me this is one of the few areas that pytorch is kind of lacking compared to other frameworks, and I'm sure this topic has been brought up in the past by other users.
In my use case I've been using torchscript + custom ops but in some cases I still find that not performant enough.

**Comment by raziel** [comment]
Yeah I understand it's a desirable treat. However, it is also not easy and perhaps not even a good idea to bundle it all in the same solution, as the complexity of a richer type system and even the semantics of "preprocessing" logic would likely hurt some use cases that care about portability but with different constraints (more stringent). That's why a layered solution would be best approach here.

> To me this is one of the few areas that pytorch is kind of lacking compared to other frameworks

I'm curious if you have any examples of frameworks, to get a sense of what an aceptable solution would look like.

Thanks

**Comment by jiyuanq** [comment]
> I'm curious if you have any examples of frameworks, to get a sense of what an aceptable solution would look like.

I think one example would be tensorflow which has a feature column API that supports commonly used light-weight feature preprocessing logic. 
There's also tf transforms which can be defined and executed independently of the model for preprocessing training data in a standalone batch job, and the resulting transforms can later loaded by model training code to be exported together with the trained model for inference.

**Comment by anijain2305** [comment]
@raziel I am marking this as low priority, since we are not working on it. But feel free to bump it.
